test/patterns/eureka/eureka.chpl

```diff
@@ -0,0 +1,93 @@
+// This is a pattern for doing parallel eureka, basically:
+//   coforall across per-locale array chunks {
+//     coforall across per-CPU array chunks {
+//       search the appropriate chunk; broadcast 'Eureka!' if we find it
+//     }
+//   }
+use BlockDist, Time;
+
+// Print timing info?
+config const printTiming = false;
+
+// This is the problem size.
+config const n = 1_000_000;
+
+// This is the array of values to search in.
+const valsD: domain(1) dmapped Block(boundingBox={0..#n}) = {0..#n};
+var vals: [valsD] int;
+[i in valsD] { vals[i] = i; }
+
+// This is the value to search for.
+config const findVal = n - 1;
+
+// If we do find findVal, this is where we'll record its index.
+const foundD: domain(1) dmapped Block(boundingBox={0..#numLocales})
+              = {0..#numLocales};
+var found: [foundD] atomic int;
+[f in found] f.write(-1); // alas, can't (yet) write this as an initial value
+
+const startTime = getCurrentTime();
+
+// Do a parallel search across the locales.
+coforall loc in Locales {
+  on loc {
+    // Figure out the part of the array stored on this locale.
+    const (lo, hi) = computeChunk(vals.domain.dim(1), here.id, numLocales);
+    const dimOnLoc = vals.domain.dim(1)(lo..hi);
+
+    // Split the on-locale work coarsely enough that every task/CPU has some.
+    const numChunks = min(dimOnLoc.length,
+                          if dataParTasksPerLocale == 0
+                          then here.maxTaskPar
+                          else dataParTasksPerLocale);
+
+    // Do a parallel search across the CPUs on this locale.
+    coforall chunk in 0..#numChunks {
+      // Do a serial search over this task's (CPU's) portion of the array.
+      const (lo, hi) = computeChunk(dimOnLoc, chunk, numChunks);
+      for i in lo..hi {
+        // If we find the value, tell everyone and quit looking.
+        if vals[i] == findVal {
+          if found[here.id].compareExchange(-1, i, memory_order_release) {
+            // The cmpxchg on found[here.id] here is a superfluous no-op.
+            [f in found] { f.compareExchange(-1, i, memory_order_release); }
+          }
+          break;
+        }
+
+        // Check now and then; if someone else found it, quit early.
+        if i % (2**17) == 0 && found[here.id].read() >= 0 {
```

Comment:
Make this frequency to check a config const?

Reply:
Would you mind if it were a config param instead of a config const?  If it's not a compile-time constant we're ultimately going to have to do an actual mod operation, but if it's a compile-time power of 2 we can just do a bit-AND.  The difference in overhead could be significant, since it happens every iteration.

Reply:
Yup, works for me -- I just want it to be configurable in some fashion. 

I don't think we (chapel compiler/modules) currently optimize `(i % 2**n) == 0` to be `(i & 2**n-1) == 0`. Is the assumption that the backend will do that?

Reply:
It's worse than I thought: making this a config const and using the same 2**17 value reduces performance by over 3x!  But if it's a config param instead, then the performance remains the same as with the explicit constant in the code.

Reply:
> Is the assumption that the backend will do that?

Yeah, I'd be very surprised if any backend (C or LLVM) failed to do so.